### PR TITLE
Add some properties relating to List.concat

### PR DESCRIPTION
**Pull request by Taneb**
I was looking for these when trying to prove that `List` is a `Monad` for `categories-examples`.

We might desire a property similar to `concat-[-]` with type `concat ∘ [_] ≗ id` but that turns out to be `++-identityʳ`.

### CODE REVIEW

File: src/Data/List/Properties.agda [hunk 1]
```diff
@@ -496,6 +496,16 @@ concat-map {f = f} xss = begin
...
 
+concat-concat : concat {a} {A} ∘ map concat ≗ concat ∘ concat
+concat-concat [] = refl
+concat-concat ([] ∷ xsss) = concat-concat xsss
+concat-concat (([] ∷ xss) ∷ xsss) = concat-concat (xss ∷ xsss)
+concat-concat (((x ∷ xs) ∷ xss) ∷ xsss) = cong (x ∷_) (concat-concat ((xs ∷ xss) ∷ xsss))
+
+concat-[-] : concat {a} {A} ∘ map [_] ≗ id
+concat-[-] [] = refl
+concat-[-] (x ∷ xs) = {!cong (x ∷_) (concat-[-] xs)!}
```
Comment (gallais) [inline src/Data/List/Properties.agda:507]: You've left a hole

File: src/Data/List/Properties.agda [hunk 2]
```diff
@@ -496,6 +496,16 @@ concat-map {f = f} xss = begin
   foldr (λ ys → map f ys ++_) [] xss         ≡⟨ sym (foldr-fusion (map f) [] (map-++-commute f) xss) ⟩
   map f (concat xss)                         ∎
 
+concat-concat : concat {a} {A} ∘ map concat ≗ concat ∘ concat
+concat-concat [] = refl
+concat-concat ([] ∷ xsss) = concat-concat xsss
+concat-concat (([] ∷ xss) ∷ xsss) = concat-concat (xss ∷ xsss)
+concat-concat (((x ∷ xs) ∷ xss) ∷ xsss) = cong (x ∷_) (concat-concat ((xs ∷ xss) ∷ xsss))
+
```
Comment (gallais) [inline src/Data/List/Properties.agda:504]: Feels like these last cases could be made cleaner by first proving
`concat-++ : concat (xs ++ ys) ≡ concat xs ++ concat ys`
and then using equational reasoning.

File: src/Data/List/Properties.agda (hunk 1 above)
Comment (Taneb) [reply src/Data/List/Properties.agda:507]: Ack, I forgot to save before committing... I'll sort that out

File: CHANGELOG.md [hunk 3]
```diff
@@ -56,4 +56,11 @@ Other minor additions
   IsLatticeIsomorphism   (⟦_⟧ : A → B) : Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂)
   ```
 
+* Added new properties to `Data.List.Properties`:
+  ```agda
+  concat-++ : concat xss ++ concat yss ≡ concat (xss ++ yss)
+  concat-concat : concat-concat : concat ∘ map concat ≗ concat ∘ concat
```
Comment (MatthewDaggitt) [inline CHANGELOG.md:62]: Repeated name on this line

File: src/Data/List/Properties.agda [hunk 4]
```diff
@@ -496,6 +496,22 @@ concat-map {f = f} xss = begin
   foldr (λ ys → map f ys ++_) [] xss         ≡⟨ sym (foldr-fusion (map f) [] (map-++-commute f) xss) ⟩
   map f (concat xss)                         ∎
 
+concat-++ : (xss yss : List (List A)) → concat xss ++ concat yss ≡ concat (xss ++ yss)
+concat-++ [] yss = refl
+concat-++ ([] ∷ xss) yss = concat-++ xss yss
+concat-++ ((x ∷ xs) ∷ xss) yss = cong (x ∷_) (concat-++ (xs ∷ xss) yss)
+
+concat-concat : concat {a} {A} ∘ map concat ≗ concat ∘ concat
```
Comment (MatthewDaggitt) [inline src/Data/List/Properties.agda:504]: I'm surprised `{a}` is necessary here, shouldn't it be inferrable from `{A}`?

File: CHANGELOG.md (hunk 3 above)
Comment (Taneb) [reply CHANGELOG.md]: Good spot, thanks

File: src/Data/List/Properties.agda (hunk 4 above)
Comment (Taneb) [reply src/Data/List/Properties.agda]: Indeed it is, I've made that `{_}` now in `concat-concat` and `concat-[-]`

Comment (MatthewDaggitt) [reply src/Data/List/Properties.agda]: The usually idiomatic way of doing this instead of having `{_} {A}` would be to instead write `{A = A}`

Comment (Taneb) [reply src/Data/List/Properties.agda]: Amended

**Pull request merged**